pkg/services/ngalert/api/api_ruler.go

```diff
@@ -122,6 +126,53 @@ func (srv RulerSrv) RouteDeleteAlertRules(c *models.ReqContext) response.Respons
 	return response.JSON(http.StatusAccepted, util.DynMap{"message": "rules deleted"})
 }
 
+// accessibleRulesFGAC will return the rules that are accessible for the user
+// while checking against FGAC. If none is accessible, an error will be returned.
+func (srv RulerSrv) accessibleRulesFGAC(rules []*ngmodels.AlertRule, hasAccess func(evaluator accesscontrol.Evaluator) bool, logger *log.ConcreteLogger) (map[string]struct{}, error) {
+	canDelete, cannotDelete := make(map[string]struct{}), make([]string, 0)
+	for _, rule := range rules {
+		if authorizeDatasourceAccessForRule(rule, hasAccess) {
+			canDelete[rule.UID] = struct{}{}
+			continue
+		}
+		cannotDelete = append(cannotDelete, rule.UID)
+	}
+
+	if len(canDelete) == 0 {
+		return nil, fmt.Errorf("%w to delete rules because user is not authorized to access data sources used by the rules", ErrAuthorization)
+	}
+
+	if len(cannotDelete) > 0 {
+		logger.Info("user cannot delete one or many alert rules because it does not have access to data sources. Those rules will be skipped", "expected", len(rules), "authorized", len(canDelete), "unauthorized", cannotDelete)
+	}
+
+	return canDelete, nil
+}
+
+// accessibleRulesProvisioning will return the rules that are accessible for
+// the user while checking against the provenance. If none is accessible, an
+// error will be returned.
+func (srv RulerSrv) accessibleRulesProvisioning(rules []*ngmodels.AlertRule, provenances map[string]ngmodels.Provenance, logger *log.ConcreteLogger) (map[string]struct{}, error) {
+	canDelete, cannotDelete := make(map[string]struct{}), make([]string, 0)
+	for _, rule := range rules {
+		if provenance, exists := provenances[rule.UID]; (exists && provenance == ngmodels.ProvenanceNone) || !exists {
+			canDelete[rule.UID] = struct{}{}
+			continue
+		}
+		cannotDelete = append(cannotDelete, rule.UID)
+	}
+
+	if len(canDelete) == 0 {
+		return nil, fmt.Errorf("all rules have been provisioned and cannot be deleted through this api")
+	}
+
+	if len(cannotDelete) > 0 {
+		logger.Info("user cannot delete one or many alert rules because it does has a provenance set. Those rules will be skipped", "expected", len(rules), "provenance_none", len(canDelete), "provenance_set", cannotDelete)
+	}
```

Comment:
The error processing here is largely duplicated in `accessibleRulesFGAC`. If we follow my other comment, this logic doesn't have to be duplicated any more.